File: workers/src/routes/sessions.ts
import { Hono } from "hono";
import type { Env, SessionRow } from "../types";
import { resolveRequestUser } from "../auth";
import { inferStatus } from "../status";

export const sessions = new Hono<{ Bindings: Env }>();

const COMPLETION_EVENT_TYPES = [
  "message.finished",
  "session.completed",
  "session.failed",
  "session.aborted",
];

function staleSeconds(env: Env): number {
  return parseInt(env.HEARTBEAT_STALE_SECONDS || "120", 10);
}

function sessionToJson(row: SessionRow) {
  return {
    session_id: row.session_id,
    session_name: row.session_name,
    session_pin: Boolean(row.session_pin),
    user_id: row.user_id,
    agent_type: row.agent_type,
    adapter_name: row.adapter_name,
    adapter_version: row.adapter_version,
    summary: row.summary,
    summary_inferred: Boolean(row.summary_inferred),
    machine_hostname: row.machine_hostname,
    machine_os: row.machine_os,
    machine_arch: row.machine_arch,
    machine_user: row.machine_user,
    machine_id: row.machine_id,
    workspace_cwd: row.workspace_cwd,
    workspace_git_branch: row.workspace_git_branch,
    workspace_project_name: row.workspace_project_name,
    current_status: row.current_status,
    latest_event_type: row.latest_event_type,
    started_at: row.started_at,
    last_event_at: row.last_event_at,
    last_heartbeat_at: row.last_heartbeat_at,
    current_run_started_at: row.current_run_started_at ?? null,
    event_count: row.event_count,
    terminal_result: row.terminal_result,
  };
}

function eventRowToJson(row: Record<string, unknown>) {
  return {
    event_id: row.event_id,
    session_id: row.session_id,
    session_name: row.session_name,
    session_pin: Boolean(row.session_pin),
    agent_type: row.agent_type,
    event_type: row.event_type,
    event_time: row.event_time,
    received_time: row.received_time,
    severity: row.severity,
    summary: row.summary,
    machine_hostname: row.machine_hostname,
    workspace_cwd: row.workspace_cwd,
    workspace_project_name: row.workspace_project_name,
    payload: row.payload_json ? JSON.parse(row.payload_json as string) : null,
  };
}

async function refreshStatuses(db: D1Database, userId: string, staleSec: number) {
  const terminal = ["completed", "failed", "aborted"];
  const result = await db
    .prepare(
      `SELECT id, latest_event_type, last_heartbeat_at, terminal_result, last_event_at, current_status
       FROM sessions WHERE user_id = ? AND current_status NOT IN (?, ?, ?)`,
    )
    .bind(userId, ...terminal)
    .all();

  if (!result.results) return;

  for (const row of result.results) {
    const r = row as Record<string, unknown>;
    const next = inferStatus(
      r.latest_event_type as string | null,
      r.last_heartbeat_at as string | null,
      r.terminal_result as string | null,
      r.last_event_at as string | null,
      staleSec,
    );
    if (next !== r.current_status) {
      await db
        .prepare(`UPDATE sessions SET current_status = ?, updated_at = datetime('now') WHERE id = ?`)
        .bind(next, r.id)
        .run();
    }
  }
}

sessions.get("/events/recent", async (c) => {
  let userId: string;
  try {
    userId = await resolveRequestUser(c.env, c.req.raw);
  } catch {
    return c.json({ detail: "Authentication required" }, 401);
  }

  await refreshStatuses(c.env.DB, userId, staleSeconds(c.env));

  const limit = Math.min(Math.max(parseInt(c.req.query("limit") || "100", 10), 1), 500);
  const completionsOnly = c.req.query("completions_only") !== "false";
  let sql = `
    SELECT e.event_id, e.session_id, e.session_name, e.session_pin, e.agent_type, e.event_type,
           e.event_time, e.received_time, e.severity, e.summary, e.machine_hostname,
           e.workspace_cwd, e.payload_json, s.workspace_project_name
    FROM events e
    LEFT JOIN sessions s ON e.session_id = s.session_id
    WHERE e.user_id = ?`;
  const params: unknown[] = [userId];

  if (completionsOnly) {
    sql += ` AND e.event_type IN (${COMPLETION_EVENT_TYPES.map(() => "?").join(", ")})`;
    params.push(...COMPLETION_EVENT_TYPES);
  }

  sql += " ORDER BY e.event_time DESC LIMIT ?";
  params.push(limit);

  const result = await c.env.DB.prepare(sql).bind(...params).all();
  return c.json({ events: (result.results || []).map((row) => eventRowToJson(row as Record<string, unknown>)) });
});

sessions.get("/sessions/live", async (c) => {
  let userId: string;
  try {
    userId = await resolveRequestUser(c.env, c.req.raw);
  } catch {
    return c.json({ detail: "Authentication required" }, 401);
  }

  await refreshStatuses(c.env.DB, userId, staleSeconds(c.env));

  const result = await c.env.DB.prepare(
    `SELECT * FROM sessions
     WHERE user_id = ? AND (current_status NOT IN ('completed', 'failed', 'aborted') OR session_pin = 1)
     ORDER BY last_event_at DESC`,
  )
    .bind(userId)
    .all<SessionRow>();

  return c.json({ sessions: (result.results || []).map(sessionToJson) });
});

sessions.get("/sessions", async (c) => {
  let userId: string;
  try {
    userId = await resolveRequestUser(c.env, c.req.raw);
  } catch {
    return c.json({ detail: "Authentication required" }, 401);
  }

  await refreshStatuses(c.env.DB, userId, staleSeconds(c.env));

  const agentType = c.req.query("agent_type");
  const status = c.req.query("status");
  const limit = Math.min(parseInt(c.req.query("limit") || "50", 10), 200);
  const offset = Math.max(parseInt(c.req.query("offset") || "0", 10), 0);

  let sql = "SELECT * FROM sessions WHERE user_id = ?";
  const params: unknown[] = [userId];

  if (agentType) {
    sql += " AND agent_type = ?";
    params.push(agentType);
  }
  if (status) {
    sql += " AND current_status = ?";
    params.push(status);
  }

  sql += " ORDER BY started_at DESC LIMIT ? OFFSET ?";
  params.push(limit, offset);

  const result = await c.env.DB.prepare(sql).bind(...params).all<SessionRow>();
  return c.json({ sessions: (result.results || []).map(sessionToJson) });
});

sessions.get("/sessions/:sessionId", async (c) => {
  let userId: string;
  try {
    userId = await resolveRequestUser(c.env, c.req.raw);
  } catch {
    return c.json({ detail: "Authentication required" }, 401);
  }

  const sessionId = c.req.param("sessionId");

  await refreshStatuses(c.env.DB, userId, staleSeconds(c.env));

  const result = await c.env.DB.prepare(
    "SELECT * FROM sessions WHERE session_id = ? AND user_id = ?",
  )
    .bind(sessionId, userId)
    .first<SessionRow>();

  if (!result) return c.json({ detail: "Session not found" }, 404);
  return c.json(sessionToJson(result));
});

sessions.delete("/sessions/:sessionId", async (c) => {
  let userId: string;
  try {
    userId = await resolveRequestUser(c.env, c.req.raw);
  } catch {
    return c.json({ detail: "Authentication required" }, 401);
  }

  const sessionId = c.req.param("sessionId");

  const session = await c.env.DB.prepare(
    "SELECT session_id FROM sessions WHERE session_id = ? AND user_id = ?",
  )
    .bind(sessionId, userId)
    .first();

  if (!session) return c.json({ detail: "Session not found" }, 404);

  await c.env.DB.prepare("DELETE FROM events WHERE session_id = ?")
    .bind(sessionId)
    .run();
  await c.env.DB.prepare("DELETE FROM sessions WHERE session_id = ?")
    .bind(sessionId)
    .run();

  return c.json({ deleted: sessionId });
});

sessions.get("/sessions/:sessionId/events", async (c) => {
  let userId: string;
  try {
    userId = await resolveRequestUser(c.env, c.req.raw);
  } catch {
    return c.json({ detail: "Authentication required" }, 401);
  }

  const sessionId = c.req.param("sessionId");
  const limit = Math.min(parseInt(c.req.query("limit") || "200", 10), 1000);

  const session = await c.env.DB.prepare(
    "SELECT user_id FROM sessions WHERE session_id = ? AND user_id = ?",
  )
    .bind(sessionId, userId)
    .first();

  if (!session) return c.json({ detail: "Session not found" }, 404);

  const result = await c.env.DB.prepare(
    `SELECT event_id, session_id, session_name, session_pin, agent_type, event_type, event_time,
            received_time, severity, summary, machine_hostname, workspace_cwd, payload_json
     FROM events WHERE session_id = ? ORDER BY event_time ASC LIMIT ?`,
  )
    .bind(sessionId, limit)
    .all();

  const events = (result.results || []).map((row) => {
    return eventRowToJson(row as Record<string, unknown>);
  });

  return c.json({ events });
});

File: server/runlight/routers/sessions.py
"""Session query endpoints."""

from __future__ import annotations

import json

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from runlight.auth import resolve_user
from runlight.db.session import get_db
from runlight.services.event_service import (
    COMPLETION_EVENT_TYPES,
    get_events_for_session,
    get_recent_events,
)
from runlight.services.session_service import (
    delete_session,
    get_all_sessions,
    get_live_sessions,
    get_session_by_id,
    refresh_session_statuses,
)

router = APIRouter(prefix="/api", tags=["sessions"])


async def _refresh_statuses(db: AsyncSession, user_id: str) -> None:
    changed = await refresh_session_statuses(db, user_id)
    if changed:
        await db.commit()


def _session_dict(s) -> dict:
    return {
        "session_id": s.session_id,
        "session_name": s.session_name,
        "session_pin": s.session_pin,
        "user_id": s.user_id,
        "agent_type": s.agent_type,
        "adapter_name": s.adapter_name,
        "adapter_version": s.adapter_version,
        "summary": s.summary,
        "summary_inferred": s.summary_inferred,
        "machine_hostname": s.machine_hostname,
        "machine_os": s.machine_os,
        "workspace_cwd": s.workspace_cwd,
        "workspace_git_branch": s.workspace_git_branch,
        "workspace_project_name": s.workspace_project_name,
        "current_status": s.current_status,
        "latest_event_type": s.latest_event_type,
        "started_at": s.started_at.isoformat() if s.started_at else None,
        "last_event_at": s.last_event_at.isoformat() if s.last_event_at else None,
        "last_heartbeat_at": s.last_heartbeat_at.isoformat() if s.last_heartbeat_at else None,
        "event_count": s.event_count,
        "terminal_result": s.terminal_result,
        "current_run_started_at": s.current_run_started_at.isoformat() if s.current_run_started_at else None,
    }


def _event_dict(e, *, workspace_project_name: str | None = None) -> dict:
    return {
        "event_id": e.event_id,
        "session_id": e.session_id,
        "session_name": e.session_name,
        "session_pin": e.session_pin,
        "agent_type": e.agent_type,
        "event_type": e.event_type,
        "event_time": e.event_time.isoformat() if e.event_time else None,
        "received_time": e.received_time.isoformat() if e.received_time else None,
        "severity": e.severity,
        "summary": e.summary,
        "machine_hostname": e.machine_hostname,
        "workspace_cwd": e.workspace_cwd,
        "workspace_project_name": workspace_project_name,
        "payload": json.loads(e.payload_json) if e.payload_json else None,
    }


@router.get("/sessions/live")
async def live_sessions(
    db: AsyncSession = Depends(get_db),
    user_id: str = Depends(resolve_user),
):
    await _refresh_statuses(db, user_id)
    sessions = await get_live_sessions(db, user_id)
    return {"sessions": [_session_dict(s) for s in sessions]}


@router.get("/sessions")
async def list_sessions(
    agent_type: str | None = Query(None),
    status: str | None = Query(None),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    user_id: str = Depends(resolve_user),
):
    await _refresh_statuses(db, user_id)
    sessions = await get_all_sessions(db, user_id, agent_type, status, limit, offset)
    return {"sessions": [_session_dict(s) for s in sessions]}


@router.get("/events/recent")
async def recent_events(
    limit: int = Query(100, ge=1, le=500),
    completions_only: bool = Query(True),
    db: AsyncSession = Depends(get_db),
    user_id: str = Depends(resolve_user),
):
    await _refresh_statuses(db, user_id)
    event_types = COMPLETION_EVENT_TYPES if completions_only else None
    rows = await get_recent_events(db, user_id, event_types, limit)
    return {"events": [_event_dict(e, workspace_project_name=pn) for e, pn in rows]}


@router.get("/sessions/{session_id}")
async def get_session(
    session_id: str,
    db: AsyncSession = Depends(get_db),
    user_id: str = Depends(resolve_user),
):
    await _refresh_statuses(db, user_id)
    session = await get_session_by_id(db, session_id)
    if session is None:
        raise HTTPException(status_code=404, detail="Session not found")
    if session.user_id != user_id:
        raise HTTPException(status_code=404, detail="Session not found")
    return _session_dict(session)


@router.delete("/sessions/{session_id}")
async def remove_session(
    session_id: str,
    db: AsyncSession = Depends(get_db),
    user_id: str = Depends(resolve_user),
):
    session = await get_session_by_id(db, session_id)
    if session is None:
        raise HTTPException(status_code=404, detail="Session not found")
    if session.user_id != user_id:
        raise HTTPException(status_code=404, detail="Session not found")
    await delete_session(db, session_id)
    await db.commit()
    return {"deleted": session_id}


@router.get("/sessions/{session_id}/events")
async def get_session_events(
    session_id: str,
    limit: int = Query(200, ge=1, le=1000),
    db: AsyncSession = Depends(get_db),
    user_id: str = Depends(resolve_user),
):
    await _refresh_statuses(db, user_id)
    session = await get_session_by_id(db, session_id)
    if session is None:
        raise HTTPException(status_code=404, detail="Session not found")
    if session.user_id != user_id:
        raise HTTPException(status_code=404, detail="Session not found")
    events = await get_events_for_session(db, session_id, limit)
    return {"events": [_event_dict(e) for e in events]}
